extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+          LOGGER.error("Failed to pull message from broker.", e);
+        }
+        finally {
+          pullRequest.getCountDownLatch().countDown();
+        }
+
+      }
+      requestsRead.clear();
+    }
+
+    /**
+     * Thread looping entry.
+     */
+    @Override
+    public void run()
+    {
+      LOGGER.info(getServiceName() + " starts.");
+      while (!isStoped()) {
+        waitForRunning(0);
+        doPull();
+      }
+
+      // in case this service is shutdown gracefully without interruption.
+      try {
+        Thread.sleep(10);
+      }
+      catch (InterruptedException e) {
+        LOGGER.error("", e);
+      }
+
+      synchronized (this) {
+        swapRequests();
+      }
+
+      doPull();
+      LOGGER.info(getServiceName() + " terminated.");
+    }
+
+    @Override
+    protected void onWaitEnd()
+    {
+      swapRequests();
+    }
+  }
+
+
+  /**
+   * Compare messages pulled from same message queue according to queue offset.
+   */
+  static final class MessageComparator implements Comparator<MessageExt>
+  {
+    @Override
+    public int compare(MessageExt lhs, MessageExt rhs)
+    {
+      return Long.compare(lhs.getQueueOffset(), lhs.getQueueOffset());
+    }
+  }
+
+
+  /**
+   * Handle message queues re-balance operations.
+   */
+  final class DruidMessageQueueListener implements MessageQueueListener
+  {
+
+    private final Set<String> topics;
+
+    private final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap;
+
+    private final DefaultMQPullConsumer defaultMQPullConsumer;
+
+    public DruidMessageQueueListener(
+        final Set<String> topics,
+        final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap,
+        final DefaultMQPullConsumer defaultMQPullConsumer
+    )
+    {
+      this.topics = topics;
+      this.topicQueueMap = topicQueueMap;
+      this.defaultMQPullConsumer = defaultMQPullConsumer;
+    }
+
+    @Override
+    public void messageQueueChanged(String topic, Set<MessageQueue> mqAll, Set<MessageQueue> mqDivided)
+    {
+      if (topics.contains(topic)) {
+        topicQueueMap.put(topic, mqDivided);
+
+        // Remove message queues that are re-assigned to other clients.
+        Iterator<ConcurrentHashMap.Entry<MessageQueue, ConcurrentSkipListSet<MessageExt>>> it =
+            messageQueueTreeSetMap.entrySet().iterator();
+        while (it.hasNext()) {
+          if (!mqDivided.contains(it.next().getKey())) {
+            it.remove();
+          }
+        }
+
+        StringBuilder stringBuilder = new StringBuilder();
+        for (MessageQueue messageQueue : mqDivided) {
+          stringBuilder.append(messageQueue.getBrokerName())
+                       .append("#")
+                       .append(messageQueue.getQueueId())
+                       .append(", ");
+        }
+
+        if (LOGGER.isDebugEnabled() && stringBuilder.length() > 2) {
+          LOGGER.debug(String.format(
```

Comment:
Yep, druid's logging wrapper is capable of doing this.  One more word, I don't think it is a good idea having this feature, we just experienced an online accident yesterday using druid which causes serious data loss and took us much time to recover and restore. The reason, after Root Cause Analysis, is part of data ingesting-filtering-transforming code directly uses logging wrapper from Druid codebase. The problem is when logging encoded URLs, the wrapper attempts to format  using String.format behind scene, but encoded, escaped urls contain %s. Boom, UnknownFormatConversionException! The point is, we do not expect an exception while logging. But it will.